rust/main/chains/hyperlane-starknet/src/merkle_tree_hook.rs

```diff
@@ -0,0 +1,198 @@
+            account,
+        );
+
+        Ok(Self {
+            contract: Arc::new(contract),
+            provider: StarknetProvider::new(locator.domain.clone(), conn),
+            conn: conn.clone(),
+        })
+    }
+
+    #[allow(unused)]
+    pub fn contract(
+        &self,
+    ) -> &StarknetMerkleTreeHookInternal<SingleOwnerAccount<AnyProvider, LocalWallet>> {
+        &self.contract
+    }
+}
+
+impl HyperlaneChain for StarknetMerkleTreeHook {
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.provider.domain()
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(self.provider.clone())
+    }
+}
+
+impl HyperlaneContract for StarknetMerkleTreeHook {
+    fn address(&self) -> H256 {
+        self.contract.address.into()
+    }
+}
+
+#[async_trait]
+impl MerkleTreeHook for StarknetMerkleTreeHook {
+    #[instrument(skip(self))]
+    async fn latest_checkpoint(&self, reorg_period: &ReorgPeriod) -> ChainResult<Checkpoint> {
+        let block_number =
+            get_block_height_for_reorg_period(&self.provider.rpc_client(), reorg_period).await?;
+
+        let (root, index) = match block_number {
+            Some(b) => self
+                .contract
+                .latest_checkpoint()
+                .block_id(starknet::core::types::BlockId::Number(b))
+                .call()
+                .await
+                .map_err(Into::<HyperlaneStarknetError>::into)?,
+            None => self
+                .contract
+                .latest_checkpoint()
+                .call()
+                .await
+                .map_err(Into::<HyperlaneStarknetError>::into)?,
+        };
+
+        Ok(Checkpoint {
+            merkle_tree_hook_address: self.address(),
+            mailbox_domain: self.domain().id(),
+            root: H256::from_slice(root.to_bytes_be().as_slice()),
+            index,
+        })
+    }
+
+    #[instrument(skip(self))]
+    #[allow(clippy::needless_range_loop)]
+    async fn tree(&self, reorg_period: &ReorgPeriod) -> ChainResult<IncrementalMerkle> {
+        let block_number =
+            get_block_height_for_reorg_period(&self.provider.rpc_client(), reorg_period).await?;
+
+        let tree = match block_number {
+            Some(b) => self
+                .contract
+                .tree()
+                .block_id(starknet::core::types::BlockId::Number(b))
+                .call()
+                .await
+                .map_err(Into::<HyperlaneStarknetError>::into)?,
+            None => self
+                .contract
+                .tree()
+                .call()
+                .await
+                .map_err(Into::<HyperlaneStarknetError>::into)?,
+        };
+
+        let mut branch = tree
+            .branch
+            .iter()
+            .map(|b| H256::from_slice(b.value.to_bytes_be().as_slice()))
+            .collect::<Vec<H256>>();
+        branch.resize(TREE_DEPTH, H256::zero());
```

Comment:
curious why this resizing was added, looking at other implementations I don't think this is done anywhere else